Require opt-in before linting a new project

novemberborn opened the pull request:
Fixes #41.

~Still needs tests and documentation.~

@gustavnikolaj if you have a chance (no rush!) please give this a go. I hope it strikes the right balance between not asking all the time yet still allowing users to opt in on a per-project or global basis. Toggle the "Reset permissions" in the package settings to reset the permissions.

novemberborn committed:
Require opt-in before linting a new project

Fixes #41.

gustavnikolaj commented:
That looks nice!

Why is the Reset Permissions option in the settings page a checkbox? Is it just because it's a work in progress, or is there something in that interaction that I dont understand? I would have expected it to be a button that you pushed to clear the permissions cache.

How is the permissions persisted between sessions?

novemberborn commented:
> Why is the Reset Permissions option in the settings page a checkbox?

Atom constructs the settings view based on a config object that is exported from the main. You can't add buttons to that view, but you can observe the settings changing. Effectively the checkbox acts as a toggle.

> How is the permissions persisted between sessions?

Atom takes care of it, see http://flight-manual.atom.io/behind-atom/sections/serialization-in-atom/. I'm not sure what happens though if you have multiple windows open at the same time. If those are separate instances of our package then presumably they'll override each other.

gustavnikolaj commented:
>> Why is the Reset Permissions option in the settings page a checkbox?
>
> Atom constructs the settings view based on a config object that is exported from the main. You can't add buttons to that view, but you can observe the settings changing. Effectively the checkbox acts as a toggle.

Could we change the UX around this, and instead make the checkbox control whether or not automatic use in all new projects are enabled?

If you press enable for "This and all future projects" we tick off the checkbox in settings too.

Resetting your list of allowed projects could be done by enabling it globally, and then disabling it again. (Which should of course be described in the settings view) Unless we could model the list of projects on the enabled-list as an array of strings in the config.

I know it is a work in progress, so forgive me if you already have plans for this. It just feels a bit awkward having a checkbox used like this.

novemberborn commented:
> Unless we could model the list of projects on the enabled-list as an array of strings in the config.

I imagine the config schema is retrieved before the package is activated, so then it can't be modified based on stored state. Also I've tried to store hashes rather than paths to avoid leaking possibly sensitive path information (which admittedly is a bit extreme).

> Resetting your list of allowed projects could be done by enabling it globally, and then disabling it again. (Which should of course be described in the settings view)

Ah that's interesting. We could have a tri state (no projects, some projects, all projects). I haven't checked what that renders like but you could reset it to "no projects" to revoke all permissions.

> I know it is a work in progress, so forgive me if you already have plans for this. It just feels a bit awkward having a checkbox used like this.

It is awkward indeed! If we go for the tri state solution, are you happy with the UX otherwise? Especially the default of not enabling all projects?

gustavnikolaj commented:
> It is awkward indeed! If we go for the tri state solution, are you happy with the UX otherwise?

It sounds like a great solution. A `<select>`-like UI would be just fine :-)

> Especially the default of not enabling all projects?

I'm fine with requiring opt-in before linting as the new default. Especially with the notification solution being as elegant as it is.

I was a bit skeptic at first, but after thinking about it for a while I actually think I understand the reasoning. This linter, given that it loads any compatible standard-engine linter and runs the code from that package, is significantly different from the standard run-of-the-mill linter-plugin. When you don't bundle the linter executable with the plugin, you cannot really assume trust.

novemberborn committed:
Test lint() with ignored files

novemberborn committed:
Remove test patterns from mocha.opts

It makes it impossible to run Mocha directly with a specific test file.

novemberborn commented:
I've implemented the tri-state solution and added tests:

![screen shot 2017-04-04 at 10 16 47](https://cloud.githubusercontent.com/assets/33538/24649695/e24a3e28-191f-11e7-9933-c53aeb06a616.png)
![screen shot 2017-04-04 at 10 16 34](https://cloud.githubusercontent.com/assets/33538/24649698/e358cbe0-191f-11e7-9676-5fcc1ec098ab.png)
![screen shot 2017-04-04 at 10 16 57](https://cloud.githubusercontent.com/assets/33538/24649705/e503ec4a-191f-11e7-9b88-552ebed68344.png)

The "Run any linter for this and all future projects" is a bit verbose, but the previous text implied only `standard` would be allowed for this and all future projects, and that's not the case.

gustavnikolaj reviewed (approved):
This looks very nice. I'll be happy to take it for a spin as soon as time permits, but if you're feeling good about it feel free to merge.
